### ISSUE TITLE
source-maps injected at runtime can't be parsed

### ISSUE THREAD

**Issue by bcoe** [GMartigny/c8CoverageTest#1]
The root of the problem is that `c8` doesn't hook `require`, like `nyc`.

A program hooking the require statement (like esm itself) is able to modify the JavaScript before it runs in the V8 engine (adding, as an example, a source-map to the bottom of the source).

[`require.extension`](https://nodejs.org/api/modules.html#modules_require_extensions) is deprecated, and can be pretty finicky, so I've been trying to see how far we can get with c8 without overriding it; however, this presents a problem. If the source-maps are inserted at runtime, and not available on disk during reporting time, it's hard for `c8` to know that they exist.

The long term solution I'd like to see, would be for Node.js and v8 to support source-maps themselves, if they see a source-map passed in to code being executed by the interpreter, it should be set aside into a cache, and available to user-land modules later on.

These same stored source-maps could be used to provide proper stack traces to programs that use transpilation.

I know @hashseed is starting to look at source-maps this year, from the point of view of V8 -- I would _love_ to help get this in motion.

**Comment by GMartigny** [comment]
Thanks for the explanation. I don't know esm or c8 that well, but could it be possible to write sourcemap to the disk (in .cache for example) and set sourcemap as an URL in the file?

**Comment by bcoe** [comment]
@GMartigny yeah, if the source-map was in `node_modules/.cache/esm`, I think we could figure something out.

I'd love to use shortcomings like this to push on better source-map support for Node.js though 😄 

CC: @jdalton

**Comment by jdalton** [comment]
How come c8 can't read inline sourcemaps?

**Comment by bcoe** [comment]
@jdalton c8 can read inline source-maps, but only if they exist in the files it's executing.

If the inline source-maps are inserted at runtime via `require.extensions` there's no way for `c8` to know about them.

Better source-map support in V8 and Node.js, I think, would be the best way to support this category of source-map, i.e., source-maps generated dynamically at runtime.

**Comment by jdalton** [comment]
Ah so c8 doesn't operate at the `vm.Script` level then. I might be able to just embed the inline source map into the cached file output .
